addons/account_asset/account_asset.py

```diff
@@ -1,351 +1,325 @@
-        for asset in self.browse(cr, uid, ids, context=context):
-            if asset.value_residual == 0.0:
-                continue
-            posted_depreciation_line_ids = depreciation_lin_obj.search(cr, uid, [('asset_id', '=', asset.id), ('move_check', '=', True)],order='depreciation_date desc')
-            old_depreciation_line_ids = depreciation_lin_obj.search(cr, uid, [('asset_id', '=', asset.id), ('move_id', '=', False)])
-            if old_depreciation_line_ids:
-                depreciation_lin_obj.unlink(cr, uid, old_depreciation_line_ids, context=context)
-
-            amount_to_depr = residual_amount = asset.value_residual
-            if asset.prorata:
-                depreciation_date = datetime.strptime(self._get_last_depreciation_date(cr, uid, [asset.id], context)[asset.id], '%Y-%m-%d')
+    @api.multi
+    def compute_depreciation_board(self):
+
+        posted_depreciation_line_ids = self.depreciation_line_ids.filtered(lambda x: x.move_check)
+        unposted_depreciation_line_ids = self.depreciation_line_ids.filtered(lambda x: not x.move_check)
+
+        # Remove old unposted depreciation lines. We cannot use unlink() with One2many field
+        commands = [(2, line_id.id, False) for line_id in unposted_depreciation_line_ids]
+
+        if self.value != 0.0:
+            amount_to_depr = residual_amount = self.value_residual
+            if self.prorata:
+                depreciation_date = datetime.strptime(self._get_last_depreciation_date()[self.id], '%Y-%m-%d')
             else:
                 # depreciation_date = 1st January of purchase year
-                purchase_date = datetime.strptime(asset.purchase_date, '%Y-%m-%d')
-                #if we already have some previous validated entries, starting date isn't 1st January but last entry + method period
-                if (len(posted_depreciation_line_ids)>0):
-                    last_depreciation_date = datetime.strptime(depreciation_lin_obj.browse(cr,uid,posted_depreciation_line_ids[0],context=context).depreciation_date, '%Y-%m-%d')
-                    depreciation_date = (last_depreciation_date+relativedelta(months=+asset.method_period))
+                asset_date = datetime.strptime(self.date, '%Y-%m-%d')
+                # if we already have some previous validated entries, starting date isn't 1st January but last entry + method period
+                if posted_depreciation_line_ids:
+                    last_depreciation_date = datetime.strptime(posted_depreciation_line_ids[0].depreciation_date, '%Y-%m-%d')
+                    depreciation_date = (last_depreciation_date+relativedelta(months=+self.method_period))
                 else:
-                    depreciation_date = datetime(purchase_date.year, 1, 1)
+                    depreciation_date = asset_date
             day = depreciation_date.day
             month = depreciation_date.month
             year = depreciation_date.year
             total_days = (year % 4) and 365 or 366
```

Comment:
This is incorrect:  `2100 % 4 ==` 0 BUT year 2100 will not have 366 days 
https://en.wikipedia.org/wiki/Leap_year#Algorithm